### DISCUSSION TITLE
SSH exec usage example

### DISCUSSION THREAD

**Discussion by devinodaniel** [Vampire/teamcity-ssh-tunnel#2] [Q&A] [answered]
Hey there. Thanks for this plugin. I'm attempting to use this plugin for an `SSH Exec` build step, but having trouble connecting. I'm getting connection timed out errors. Likely because I'm setting it up incorrectly.

Would you be able to provide an example of what settings I should use in the SSH Exec if I'm trying to get TeamCity to run an SSH command on a remote host, proxied through a different remote host?

I'm mostly unsure of what I should put in the **Deployment Target** section for `Target` and `Port`.

I've tried:

Target: %sshTunnel.<sanitized_name>.remote.address%
Port: %sshTunnel.<sanitized_name>.remote.port%

But no luck. I have confirmed access from the proxy works as expected.

**Comment by Vampire** [comment]
You are aware that `<sanitized_name>` is a placeholder and that TC shows you all such properties you can use with your current config as list or also with auto-completion?

**Comment by devinodaniel** [comment]
Yes, of course. I removed the name I have there, as it's irrelevant.

**Comment by Vampire** [comment, marked as answer]
Ok, just wanted to rule out the obvious first. :-)
So, to be able to answer reliably I would need more information about your config.
But I just try to guess your setup.
It is like this I guess?
![grafik](https://user-images.githubusercontent.com/325196/67815898-88517400-faa8-11e9-81c7-93196db81aa3.png)

In that case you would configure the SSH tunnel build feature like
![grafik](https://user-images.githubusercontent.com/325196/67816138-4d037500-faa9-11e9-84e1-c5e34031cdf5.png)

And the SSH Exec build step like
![grafik](https://user-images.githubusercontent.com/325196/67816596-bafc6c00-faaa-11e9-8813-0d90ada76bd4.png)

If you could use the `remote.address` and `remote.port` directly, you wouldn't need an SSH tunnel through the proxy, that's the whole point of an SSH tunnel.
You connect to the `local.address` and `local.port`, the SSH client that is listening there, forwards the connection via the SSH connection to proxy and connects to `remote.address` and `remote.port` from there as proxy is allowed to connect to target while TC is not.

**Comment by devinodaniel** [comment]
Thank you for the detailed explanation, this is helpful. I tested and it does work. I never thought to use `local.address` and `local.port` because I assumed that TC needed to know the final destination in the SSH Exec settings. But it makes sense it would automatically forward from the local. Very nice! Thanks again for this addition to TC!
